components/script/dom/gpuconvert.rs

```diff
@@ -409,68 +418,78 @@ pub fn convert_store_op(op: Option<GPUStoreOp>) -> wgpu_com::StoreOp {
-                ))
-            } else {
-                Ok(wgt::Origin3d {
-                    x: v.get(0).copied().unwrap_or(0),
-                    y: v.get(1).copied().unwrap_or(0),
-                    z: v.get(2).copied().unwrap_or(0),
-                })
-            }
-        },
-        GPUOrigin3D::GPUOrigin3DDict(d) => Ok(wgt::Origin3d {
-            x: d.x,
-            y: d.y,
-            z: d.z,
-        }),
+impl TryFrom<&GPUOrigin3D> for wgt::Origin3d {
+    type Error = Error;
+
+    fn try_from(origin: &GPUOrigin3D) -> Result<Self, Self::Error> {
+        match origin {
+            GPUOrigin3D::RangeEnforcedUnsignedLongSequence(v) => {
+                // https://gpuweb.github.io/gpuweb/#abstract-opdef-validate-gpuorigin3d-shape
+                if v.len() > 3 {
+                    Err(Error::Type(
+                        "sequence is too long for GPUOrigin3D".to_string(),
+                    ))
+                } else {
+                    Ok(wgt::Origin3d {
+                        x: v.get(0).copied().unwrap_or(0),
+                        y: v.get(1).copied().unwrap_or(0),
+                        z: v.get(2).copied().unwrap_or(0),
+                    })
+                }
+            },
+            GPUOrigin3D::GPUOrigin3DDict(d) => Ok(wgt::Origin3d {
+                x: d.x,
+                y: d.y,
+                z: d.z,
+            }),
+        }
     }
 }
 
-pub fn convert_ic_texture(
-    ic_texture: &GPUImageCopyTexture,
-) -> Fallible<wgpu_com::ImageCopyTexture> {
-    Ok(wgpu_com::ImageCopyTexture {
-        texture: ic_texture.texture.id().0,
-        mip_level: ic_texture.mipLevel,
-        origin: ic_texture
-            .origin
-            .as_ref()
-            .map(|origin| convert_origin3d(origin))
-            .transpose()?
-            .unwrap_or_default(),
-        aspect: match ic_texture.aspect {
-            GPUTextureAspect::All => wgt::TextureAspect::All,
-            GPUTextureAspect::Stencil_only => wgt::TextureAspect::StencilOnly,
-            GPUTextureAspect::Depth_only => wgt::TextureAspect::DepthOnly,
-        },
-    })
+impl TryFrom<&GPUImageCopyTexture> for wgpu_com::ImageCopyTexture {
+    type Error = Error;
+
+    fn try_from(ic_texture: &GPUImageCopyTexture) -> Result<Self, Self::Error> {
+        Ok(wgpu_com::ImageCopyTexture {
+            texture: ic_texture.texture.id().0,
+            mip_level: ic_texture.mipLevel,
+            origin: ic_texture
+                .origin
+                .as_ref()
+                .map(|origin| wgt::Origin3d::try_from(origin))
+                .transpose()?
+                .unwrap_or_default(),
+            aspect: match ic_texture.aspect {
+                GPUTextureAspect::All => wgt::TextureAspect::All,
+                GPUTextureAspect::Stencil_only => wgt::TextureAspect::StencilOnly,
+                GPUTextureAspect::Depth_only => wgt::TextureAspect::DepthOnly,
+            },
+        })
+    }
 }
 
```

Comment:
@sagudev could not convert some function because types are defined outside this crate, like `convert_bind_group_layout_entry`.

Reply:
I think in such cases it should be still possible to implement [`Into`](https://doc.rust-lang.org/std/convert/trait.Into.html), but we can do that in a followup.